### ISSUE TITLE
Steam must be running to use this feature.......

### ISSUE THREAD

**Issue by Freejack01** [aclist/dztui#76]
Steam is open and running, this has been an on and off issue since the last update. Sometimes it tells me the API is not available. Also it was working great on local lan going out to my internet IP and getting the modlist. After the last update it now refuses to get the modlist of which it worked just fine to this recent update.

I appreciate all your hard work.

Current Version is 3.3.5 the version I know that worked and worked well was 3.3.2.

**Comment by aclist** [comment]
Thanks for your message. Although it may appear that an older version of DZGUI "worked better," please consider that there are also external, architectural changes that happen with Steam itself over time, making this a false positive.

Specifically, this is not related to changes in DZGUI itself, but to a host of bugs going on with Steam. Steam's latest update introduced numerous changes that break functionality with different systems and with third party programs. This is because the Steam client is using a new UI that was formerly considered the beta UI.

I will need to refactor some parts of DZGUI to properly target the new Steam client, but Steam is functionally broken for a lot of users right now on different distros and window manager combinations, so it is difficult to test. If you are interested in a snapshot of the situation, you can look at Valve's steam-for-linux issue tracker and you will see several dozen bug reports.

Getting down to business: 
1. I think I can drop the "is Steam running" check for now until the situation stabilizes if that would help.
2. Third party APIs may have sporadic downtime from time to time. Notably, Steam has scheduled downtime on some of its interface every Tuesday, so this may contribute to periodic outages and degradation of service. How frequent would you say this is?
3. If you can provide the public IP of your server, I can test the third point. If you would rather share it privately, please provide some point of contact for doing so. Looking over the list of changes, I don't see any refactoring of code that would adversely affect polling servers, so its possible your server fell off the public index for some reason.

**Comment by Freejack01** [comment]
The Api Issue has been at least 2 times a day for the last 2 weeks, the ip is 24.169.105.46. With that said what part of the world would you be coming from? I have a heavily indentured firewall that only likes certain places. I can open it up for our testing purposes.

**Comment by aclist** [comment]
Thanks for the information. I should clarify that DZGUI does not query servers directly, so no firewall workaround needed. I don't need to connect directly to test. DZGUI queries the Valve master server for an index of servers and passes that metadata on to the user. It's not possible to query DayZ servers directly and receive a well-formed modlist response due to the fact that Bohemia changed the way they encode and stuff responses into the A2S_RULES protocol, and this has not been sufficiently explained to anyone at this point in time. Some maintainers of other projects have attempted to reverse engineer this, but the results are incomplete and not reliable.

Anyway, I had a look at that IP and some other IPs that should be returning responses correctly. What I found is that (as of this writing), the Valve interface was not responding correctly to queries and returning an empty response, regardless of IP. So I believe this is affecting more than your server specifically, but the entire Connect By IP method. This interface seems to periodically go down. I will check it again later, but it seems like it's not reliable in the long term. I cannot say whether the interface will be down indefinitely.

As an alternative, we can query the entire index of servers and then search for the IP you asked for. This would have the same effect and still prompt you to connect by IP, but the query behind the scenes would be different and shouldn't affect performance aside from a slight bump in parsing the list. I tried this approach with your IP and it found it correctly. It took about 15s. To compensate for this, I have added a progress dialog to this step.

I have a fix worked up, but this also affects other parts of the application like the "Recent servers" menu, which relies on the same method, so it is highly experimental and I am not ready to release this patch yet. So far, I am able to at least find your server in the index and get the query port.

The problem is twofold, though. It looks like the API responsible for querying your server's modlist is timing out, and I can't say for certain why that is, since DZGUI does not sit in between the connection between the API server and your server. Are you allowing queries on port 27016? It looks like Battlemetrics at least indexed you, but there are other trackers that may be trying to query the server in an ad-hoc fashion for the A2S_RULES response as specified [here](https://developer.valvesoftware.com/wiki/Server_queries) and timing out.

Lastly, an important question: is this server running DayZ Experimental or DayZ Vanilla?

**Comment by Freejack01** [comment]
The server is Dayz Vanilla, which API is timing out? It was working without issue I guess up to this recent Steam update.

**Comment by aclist** [comment]
As mentioned, our preferred method is not working:
- [Valve's interface](https://partner.steamgames.com/doc/webapi/IGameServersService) knows of the existence of your server but cannot query it specifically

And anecdotally, as a test, 
- [This third party tool](https://daemonforge.dev/Server/) returns a network error
- [This other third party tool](https://dayzsalauncher.com/#/servercheck/) returns a network error

BattleMetrics doesn't do a live check AFAIK, just periodically builds a manifest and prints the stored data.

To isolate unrelated variables, probably best to start with a barebones config on your server and see if it can reliably be probed by any external tool leveraging A2S_RULES.

I can't point to a single thing right now, as there are quite a lot of unrelated parameters here, such as Valve endpoint connectivity, changes to Steam itself, your network settings, so would be good to establish a control here.

I did push a small hotfix for the initial issue, so you should at least be able to run the mod installation process without being nagged about Steam not being running.

**Comment by Freejack01** [comment]
Also the force update mods in advanced options with the console no longer works also. As of now I can launch and run DayZ without it telling me Steam is not running. I figured out the server issue now that I can run the client. It appears it was a firewall issue where something changed and I had to reconfigure the setup to work.

**Comment by aclist** [comment]
Thank you, could you elaborate on "no longer works?"

 - Is it not raising the Steam window?
 - Is it raising the Steam window but not passing the right mod commands?
 - Is it staging the mods for download but not proceeding to download?
 - Is the option failing altogether and dumping you back to the main menu of DZGUI?

**Comment by Freejack01** [comment]
It opens the new console window and does not download any of the mods and at the top of the steam window it says:

Cannot read property "GetAccountID" of null

**Comment by aclist** [comment]
New Steam UI or legacy UI? (-vgui)

**Comment by aclist** [comment]
About the error message itself, it is not caused by DZGUI. Steam console may periodically report informational messages into the console regardless of user input, and these warnings can be ignored.

As for the targeting, I investigated it, and it looks like the window ID taxonomy has changed, as the Steam client reports itself to the window manager as `steamwebhelper.steam` now. 

I have made a small fix that should properly request files.

Note that force updating local mods does not trigger every mod to be redownloaded. So if you see small download progress updates that finish instantly, this is correct.

When you request downloads, an internal integrity check is performed. Only if any of the files within a mod has materially changed from the upstream version will it fetch the necessary updates. For example, if a single file was deleted from a mod, force updating it will notice this change and download the missing file. If the contents of a file were changed, similarly it will be downloaded. This is an atomic operation that means force updating a mod does not require redownloadin tons of data.

Also note that this feature is still experimental and is intended for power users.
